src/components/AccountSeed.js

```diff
@@ -17,154 +17,93 @@
-  }
-
-  getSuggestions(input, words) {
-    let fromIndex = WORDS.findIndex(w => w.startsWith(input));
+  getSuggestions = (inputWordArray, word_list) => {
+    const input = inputWordArray[inputWordArray.length - 1]
+    // the word list is sorted, get the index we should start searching
+    // for the word
+    let fromIndex = word_list.findIndex(w => w.startsWith(input));
     if (fromIndex === -1) {
       return [];
     }
-    if (WORDS[fromIndex] === input) {
-      fromIndex = 0;
-    }
+
     const SUGGESTIONS_COUNT = 5;
     const result = [];
     let yielded = 0;
-    result.push(WORDS[fromIndex]);
-    while (yielded < SUGGESTIONS_COUNT && WORDS[fromIndex] !== undefined) {
-      ++fromIndex;
-      if (!WORDS[fromIndex].startsWith(input)) {
+    while (yielded < SUGGESTIONS_COUNT && fromIndex < word_list.length) {
+      if (!word_list[fromIndex].startsWith(input)) {
         return result;
       }
-      if (words.indexOf(WORDS[fromIndex]) !== -1) {
-        continue;
+
+      // do not suggest words that where already used (and part of `inputWordArray`)
+      // unless it's the latest word e.g what the user is typing
+      if (inputWordArray.indexOf(word_list[fromIndex]) === -1 || inputWordArray.indexOf(word_list[fromIndex]) === (inputWordArray.length - 1)) {
+        result.push(word_list[fromIndex]);
+        yielded++
       }
-      result.push(WORDS[fromIndex]);
-      ++yielded;
+      fromIndex++;
     }
     return result;
   }
 
-  renderSuggestions() {
-    const { value } = this.props;
-    const words = value.length ? value.split(' ') : [];
-    const wordPosition = this.getWordPosition();
-    let searchInput = this.getSearchInput();
-    if (WORDS_INDEX[searchInput]) {
-      searchInput = '';
+  narrowWordList = (inputWordArray) => {
+    // if the last word is only present in BIP39 word list
+    if (BIP39_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) !== -1 && PARITY_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) === -1) {
+      this.setState({ useParityWordList: false })
+      // if the second last word is only present in Parity word list
+    } else if (BIP39_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) === -1 && PARITY_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) !== -1) {
+      this.setState({ useBIP39WordList: false })
     }
-    const suggestions = this.getSuggestions(searchInput, words);
+  }
+
+  generateSuggestions = (inputWordArray) => {
+    const { useBIP39WordList, useParityWordList } = this.state;
+
+    // try to narrow down the word list using the last word typed, 
+    // as soon as a second word is being typed
+    inputWordArray.length > 1 && useBIP39WordList && useParityWordList && this.narrowWordList(inputWordArray);
+
+    let suggestions = []
+    useParityWordList && suggestions.push(...this.getSuggestions(inputWordArray, PARITY_WORDS));
+    useBIP39WordList && suggestions.push(...this.getSuggestions(inputWordArray, BIP39_WORDS));
```

Comment:
`push` and spread is a total overkill here, plus I hate using logical and as shortand `if`.

```js
let suggestions = [];

if (useParityWordList) {
  suggestions = suggestions.concat(this.getSuggestions(inputWordArray, PARITY_WORDS));
}
if (useBIP39WordList) {
  suggestions = suggestions.concat(this.getSuggestions(inputWordArray, BIP39_WORDS));
}
```

If we really want a one-liner, I'd prefer `if (foo) bar;` to `foo && bar;` as it makes intent clearer, but I tend to err on having mandatory braces anyway.

Reply:
Just for my information, is what you suggest performing better, or is that just preferences? I prefer `&&` compared to `if` but I can change my preferences :)

Reply:
I don't think it performs better, it's just less magical and more obvious to whoever is going to read it later what it is you wanted to do. Replacing `if` with `&&` is the sort of a thing `UglifyJS` does to reduce file size :P.